File: apps/desktop-electron/renderer/src/tokens/themeDerivation.ts
import type { CustomThemeColors } from "./customThemes";

export type ThemeTokenMap = Record<string, string>;

const TOKEN_KEYS = [
  "bg-app",
  "bg-panel",
  "bg-panel-alt",
  "bg-input",
  "bg-toolbar",
  "border-highlight",
  "border-shadow",
  "border-outline",
  "border-light",
  "border-muted",
  "bevel-highlight",
  "bevel-shadow",
  "fg-primary",
  "fg-muted",
  "fg-disabled",
  "fg-inverse",
  "accent-primary",
  "accent-primary-hover",
  "accent-primary-pressed",
  "accent-secondary",
  "accent-secondary-hover",
  "accent-success",
  "accent-warning",
  "accent-danger",
  "accent-info",
  "overlay-header",
  "overlay-backdrop",
  "row-unmatched-bg",
] as const;

export const DERIVED_THEME_TOKEN_KEYS = TOKEN_KEYS;

function clamp(n: number, min = 0, max = 255): number {
  return Math.min(max, Math.max(min, n));
}

export function normalizeHex(hex: string): string {
  const h = hex.trim().replace(/^#/, "");
  if (h.length === 3) {
    return `#${h[0]}${h[0]}${h[1]}${h[1]}${h[2]}${h[2]}`.toLowerCase();
  }
  if (h.length === 6) return `#${h}`.toLowerCase();
  throw new Error(`Invalid hex color: ${hex}`);
}

function hexToRgb(hex: string): { r: number; g: number; b: number } {
  const n = normalizeHex(hex).slice(1);
  return {
    r: parseInt(n.slice(0, 2), 16),
    g: parseInt(n.slice(2, 4), 16),
    b: parseInt(n.slice(4, 6), 16),
  };
}

function rgbToHex(r: number, g: number, b: number): string {
  return `#${[r, g, b].map((v) => clamp(Math.round(v)).toString(16).padStart(2, "0")).join("")}`;
}

export function darken(hex: string, amount: number): string {
  const { r, g, b } = hexToRgb(hex);
  const f = 1 - amount;
  return rgbToHex(r * f, g * f, b * f);
}

export function lighten(hex: string, amount: number): string {
  const { r, g, b } = hexToRgb(hex);
  return rgbToHex(r + (255 - r) * amount, g + (255 - g) * amount, b + (255 - b) * amount);
}

function relativeLuminance(hex: string): number {
  const { r, g, b } = hexToRgb(hex);
  const [rs, gs, bs] = [r, g, b].map((c) => {
    const s = c / 255;
    return s <= 0.03928 ? s / 12.92 : ((s + 0.055) / 1.055) ** 2.4;
  });
  return 0.2126 * rs + 0.7152 * gs + 0.0722 * bs;
}

export function pickContrastText(bgHex: string): string {
  return relativeLuminance(bgHex) > 0.45 ? "#000000" : "#ffffff";
}

export function deriveThemeTokens(colors: CustomThemeColors): ThemeTokenMap {
  const bgApp = normalizeHex(colors.bgApp);
  const bgPanel = normalizeHex(colors.bgPanel);
  const bgInput = normalizeHex(colors.bgInput);
  const fgPrimary = normalizeHex(colors.fgPrimary);
  const fgMuted = normalizeHex(colors.fgMuted);
  const accentPrimary = normalizeHex(colors.accentPrimary);

  const bgToolbar = darken(bgApp, 0.05);
  const bgPanelAlt = darken(bgPanel, 0.08);
  const borderShadow = darken(bgApp, 0.2);
  const borderHighlight = lighten(bgPanel, 0.15);
  const borderLight = lighten(bgPanel, 0.22);
  const borderMuted = darken(bgPanel, 0.12);

  return {
    "bg-app": bgApp,
    "bg-panel": bgPanel,
    "bg-panel-alt": bgPanelAlt,
    "bg-input": bgInput,
    "bg-toolbar": bgToolbar,
    "border-highlight": borderHighlight,
    "border-shadow": borderShadow,
    "border-outline": "#000000",
    "border-light": borderLight,
    "border-muted": borderMuted,
    "bevel-highlight": borderHighlight,
    "bevel-shadow": borderShadow,
    "fg-primary": fgPrimary,
    "fg-muted": fgMuted,
    "fg-disabled": fgMuted,
    "fg-inverse": pickContrastText(accentPrimary),
    "accent-primary": accentPrimary,
    "accent-primary-hover": lighten(accentPrimary, 0.12),
    "accent-primary-pressed": darken(accentPrimary, 0.12),
    "accent-secondary": bgPanelAlt,
    "accent-secondary-hover": lighten(bgPanelAlt, 0.08),
    "accent-success": normalizeHex(colors.accentSuccess),
    "accent-warning": normalizeHex(colors.accentWarning),
    "accent-danger": normalizeHex(colors.accentDanger),
    "accent-info": lighten(accentPrimary, 0.05),
    "overlay-header": "rgba(0, 0, 0, 0.25)",
    "overlay-backdrop": "rgba(0, 0, 0, 0.72)",
    "row-unmatched-bg": darken(normalizeHex(colors.accentDanger), 0.35),
  };
}

export function applyThemeTokensToDocument(tokens: ThemeTokenMap): void {
  const root = document.documentElement;
  for (const key of DERIVED_THEME_TOKEN_KEYS) {
    const value = tokens[key];
    if (value) root.style.setProperty(`--${key}`, value);
  }
}

export function clearThemeTokenOverrides(): void {
  const root = document.documentElement;
  for (const key of DERIVED_THEME_TOKEN_KEYS) {
    root.style.removeProperty(`--${key}`);
  }
}

File: archive/legacy-pyside6-ui/appearance/theme_derivation.py
"""Derive full theme tokens from eight editor colors (lab parity)."""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Dict

ThemeTokenMap = Dict[str, str]

_HEX_RE = re.compile(r"^#?([0-9a-fA-F]{3}|[0-9a-fA-F]{6})$")


@dataclass(frozen=True)
class CustomThemeColors:
    bg_app: str
    bg_panel: str
    bg_input: str
    fg_primary: str
    fg_muted: str
    accent_primary: str
    accent_success: str
    accent_warning: str
    accent_danger: str


@dataclass
class CustomTheme:
    id: str
    name: str
    colors: CustomThemeColors


def _clamp(value: float, low: float = 0.0, high: float = 255.0) -> int:
    return int(min(high, max(low, round(value))))


def normalize_hex(hex_color: str) -> str:
    raw = hex_color.strip()
    if not _HEX_RE.match(raw):
        raise ValueError(f"Invalid hex color: {hex_color}")
    body = raw.lstrip("#")
    if len(body) == 3:
        body = "".join(ch * 2 for ch in body)
    return f"#{body.lower()}"


def _hex_to_rgb(hex_color: str) -> tuple[int, int, int]:
    normalized = normalize_hex(hex_color).lstrip("#")
    return (
        int(normalized[0:2], 16),
        int(normalized[2:4], 16),
        int(normalized[4:6], 16),
    )


def _rgb_to_hex(r: int, g: int, b: int) -> str:
    return f"#{_clamp(r):02x}{_clamp(g):02x}{_clamp(b):02x}"


def darken(hex_color: str, amount: float) -> str:
    r, g, b = _hex_to_rgb(hex_color)
    factor = 1.0 - amount
    return _rgb_to_hex(r * factor, g * factor, b * factor)


def lighten(hex_color: str, amount: float) -> str:
    r, g, b = _hex_to_rgb(hex_color)
    return _rgb_to_hex(
        r + (255 - r) * amount,
        g + (255 - g) * amount,
        b + (255 - b) * amount,
    )


def _relative_luminance(hex_color: str) -> float:
    r, g, b = _hex_to_rgb(hex_color)

    def channel(value: int) -> float:
        s = value / 255.0
        if s <= 0.03928:
            return s / 12.92
        return ((s + 0.055) / 1.055) ** 2.4

    rs, gs, bs = channel(r), channel(g), channel(b)
    return 0.2126 * rs + 0.7152 * gs + 0.0722 * bs


def pick_contrast_text(bg_hex: str) -> str:
    return "#000000" if _relative_luminance(bg_hex) > 0.45 else "#ffffff"


def derive_theme_tokens(colors: CustomThemeColors) -> ThemeTokenMap:
    bg_app = normalize_hex(colors.bg_app)
    bg_panel = normalize_hex(colors.bg_panel)
    bg_input = normalize_hex(colors.bg_input)
    fg_primary = normalize_hex(colors.fg_primary)
    fg_muted = normalize_hex(colors.fg_muted)
    accent_primary = normalize_hex(colors.accent_primary)

    bg_toolbar = darken(bg_app, 0.05)
    bg_panel_alt = darken(bg_panel, 0.08)
    border_shadow = darken(bg_app, 0.2)
    border_highlight = lighten(bg_panel, 0.15)
    border_light = lighten(bg_panel, 0.22)
    border_muted = darken(bg_panel, 0.12)

    return {
        "bg-app": bg_app,
        "bg-panel": bg_panel,
        "bg-panel-alt": bg_panel_alt,
        "bg-input": bg_input,
        "bg-toolbar": bg_toolbar,
        "border-highlight": border_highlight,
        "border-shadow": border_shadow,
        "border-outline": "#000000",
        "border-light": border_light,
        "border-muted": border_muted,
        "bevel-highlight": border_highlight,
        "bevel-shadow": border_shadow,
        "fg-primary": fg_primary,
        "fg-muted": fg_muted,
        "fg-disabled": fg_muted,
        "fg-inverse": pick_contrast_text(accent_primary),
        "accent-primary": accent_primary,
        "accent-primary-hover": lighten(accent_primary, 0.12),
        "accent-primary-pressed": darken(accent_primary, 0.12),
        "accent-secondary": bg_panel_alt,
        "accent-secondary-hover": lighten(bg_panel_alt, 0.08),
        "accent-success": normalize_hex(colors.accent_success),
        "accent-warning": normalize_hex(colors.accent_warning),
        "accent-danger": normalize_hex(colors.accent_danger),
        "accent-info": lighten(accent_primary, 0.05),
        "overlay-header": "rgba(0, 0, 0, 0.25)",
        "overlay-backdrop": "rgba(0, 0, 0, 0.72)",
    }
